pkg/reconciler/build/build.go

```diff
@@ -109,29 +125,151 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	if err = c.validateBuild(build); err != nil {
+		c.Logger.Errorf("Failed to validate build: %v", err)
+		return err
+	}
+
+	// If the build's done, then ignore it.
+	if !builder.IsDone(&build.Status) {
+		// If the build is not done, but is in progress (has an operation), then asynchronously wait for it.
+		// TODO(mattmoor): Check whether the Builder matches the kind of our c.builder.
+		if build.Status.Builder != "" {
+			op, err := c.builder.OperationFromStatus(&build.Status)
+			if err != nil {
+				return err
+			}
+
+			// Check if build has timed out
+			if builder.IsTimeout(&build.Status, build.Spec.Timeout) {
+				//cleanup operation and update status
+				timeoutMsg := fmt.Sprintf("Build %q failed to finish within %q", build.Name, build.Spec.Timeout.Duration.String())
+
+				if err := op.Terminate(); err != nil {
+					c.Logger.Errorf("Failed to terminate pod: %v", err)
+					return err
+				}
+
+				build.Status.SetCondition(&duckv1alpha1.Condition{
+					Type:    v1alpha1.BuildSucceeded,
+					Status:  corev1.ConditionFalse,
+					Reason:  "BuildTimeout",
+					Message: timeoutMsg,
+				})
+				// update build completed time
+				build.Status.CompletionTime = metav1.Now()
+
+				if _, err := c.updateStatus(build); err != nil {
+					c.Logger.Errorf("Failed to update status for pod: %v", err)
+					return err
+				}
 
+				c.Logger.Errorf("Timeout: %v", timeoutMsg)
+				return nil
+			}
+
+			// if not timed out then wait async
+			go c.waitForOperation(build, op)
+		} else {
+			build.Status.Builder = c.builder.Builder()
+			// If the build hasn't even started, then start it and record the operation in our status.
+			// Note that by recording our status, we will trigger a reconciliation, so the wait above
+			// will kick in.
+			var tmpl v1alpha1.BuildTemplateInterface
+			if build.Spec.Template != nil {
+				if build.Spec.Template.Kind == v1alpha1.ClusterBuildTemplateKind {
+					tmpl, err = c.clusterBuildTemplatesLister.Get(build.Spec.Template.Name)
+					if err != nil {
+						// The ClusterBuildTemplate resource may not exist.
+						if errors.IsNotFound(err) {
+							runtime.HandleError(fmt.Errorf("cluster build template %q does not exist", build.Spec.Template.Name))
+						}
+						return err
+					}
+				} else {
+					tmpl, err = c.buildTemplatesLister.BuildTemplates(namespace).Get(build.Spec.Template.Name)
+					if err != nil {
+						// The BuildTemplate resource may not exist.
+						if errors.IsNotFound(err) {
+							runtime.HandleError(fmt.Errorf("build template %q in namespace %q does not exist", build.Spec.Template.Name, namespace))
+						}
+						return err
+					}
+				}
+			}
+			build, err = builder.ApplyTemplate(build, tmpl)
+			if err != nil {
+				return err
+			}
+			// TODO: Validate build except steps+template
+			b, err := c.builder.BuildFromSpec(build)
+			if err != nil {
+				return err
+			}
+			op, err := b.Execute()
+			if err != nil {
+				build.Status.SetCondition(&duckv1alpha1.Condition{
```

Comment:
Another thing to pick up is the use of higher-level functions (I actually removed `setCondition` from our types).

Reply:
Yeah, that's definitely a good idea. Probably easier to do once we remove the builder split, since some `SetCondition` calls are in there. It would be much easier to make this change if everything was linear.

My goal with this PR was to change as little code as possible, even though there are plenty of threads to pull. But getting this in before 0.2 is the priority I think. More cleanups next week! 🎉